src/scheduler/gc_works.rs

```diff
@@ -0,0 +1,462 @@
+use super::*;
+use crate::plan::global::GcStatus;
+use crate::util::*;
+use crate::vm::*;
+use crate::*;
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Mutex;
+
+pub struct ScheduleCollection;
+
+unsafe impl Sync for ScheduleCollection {}
+
+impl<VM: VMBinding> GCWork<VM> for ScheduleCollection {
+    fn do_work(&mut self, worker: &mut GCWorker<VM>, mmtk: &'static MMTK<VM>) {
+        mmtk.plan.schedule_collection(worker.scheduler());
+    }
+}
+
+impl<VM: VMBinding> CoordinatorWork<MMTK<VM>> for ScheduleCollection {}
+
+/// GC Preparation Work (include updating global states)
+pub struct Prepare<P: Plan> {
+    pub plan: &'static P,
+}
+
+unsafe impl<P: Plan> Sync for Prepare<P> {}
+
+impl<P: Plan> Prepare<P> {
+    pub fn new(plan: &'static P) -> Self {
+        Self { plan }
+    }
+}
+
+impl<P: Plan> GCWork<P::VM> for Prepare<P> {
+    fn do_work(&mut self, worker: &mut GCWorker<P::VM>, mmtk: &'static MMTK<P::VM>) {
+        trace!("Prepare Global");
+        self.plan.prepare(worker.tls);
+        let _guard = MUTATOR_ITERATOR_LOCK.lock().unwrap();
+        for mutator in <P::VM as VMBinding>::VMActivePlan::mutators() {
+            let mutator = unsafe { &mut *(mutator as *mut _ as *mut P::Mutator) };
+            mmtk.scheduler
+                .prepare_stage
+                .add(PrepareMutator::<P>::new(self.plan, mutator));
+        }
+        for w in &worker.group().unwrap().workers {
+            w.local_works.add(PrepareCollector::default());
+        }
+    }
+}
+
+/// GC Preparation Work (include updating global states)
+pub struct PrepareMutator<P: Plan> {
+    pub plan: &'static P,
+    pub mutator: &'static mut P::Mutator,
```

Comment:
Yes, it's correct because this work-packet does not outlive a GC. I've added a comment here to explain this.

Probably we need a `<'gc>` lifetime parameter in the future?